File: crop-me/src/main/java/com/mikelau/croperino/BitmapManager.java
package com.mikelau.croperino;

import android.graphics.BitmapFactory;

import java.util.Iterator;
import java.util.WeakHashMap;

public class BitmapManager {

    private final WeakHashMap<Thread, ThreadStatus> mThreadStatus = new WeakHashMap<>();
    private static BitmapManager sManager = null;
    private enum State {CANCEL, ALLOW}

    private BitmapManager() {}

    private static class ThreadStatus {
        public State mState = State.ALLOW;
        public BitmapFactory.Options mOptions;

        @Override
        public String toString() {
            String s;
            if (mState == State.CANCEL) {
                s = "Cancel";
            } else if (mState == State.ALLOW) {
                s = "Allow";
            } else {
                s = "?";
            }
            s = "thread state = " + s + ", options = " + mOptions;
            return s;
        }
    }

    public static class ThreadSet implements Iterable<Thread> {
        private final WeakHashMap<Thread, Object> mWeakCollection = new WeakHashMap<>();

        public void add(Thread t) {
            mWeakCollection.put(t, null);
        }

        public Iterator<Thread> iterator() {
            return mWeakCollection.keySet().iterator();
        }
    }

    private synchronized ThreadStatus getOrCreateThreadStatus(Thread t) {
        ThreadStatus status = mThreadStatus.get(t);
        if (status == null) {
            status = new ThreadStatus();
            mThreadStatus.put(t, status);
        }
        return status;
    }

    public synchronized void cancelThreadDecoding(ThreadSet threads) {
        for (Thread t : threads) {
            cancelThreadDecoding(t);
        }
    }

    public synchronized void cancelThreadDecoding(Thread t) {
        ThreadStatus status = getOrCreateThreadStatus(t);
        status.mState = State.CANCEL;
        if (status.mOptions != null) {
            status.mOptions.requestCancelDecode();
        }

        notifyAll();
    }

    public static synchronized BitmapManager instance() {
        if (sManager == null) {
            sManager = new BitmapManager();
        }
        return sManager;
    }
}

File: croperino/src/main/java/com/mikelau/croperino/BitmapManager.kt
package com.mikelau.croperino

import android.graphics.BitmapFactory
import java.util.WeakHashMap

class BitmapManager private constructor() {
    private val mThreadStatus = WeakHashMap<Thread, ThreadStatus>()

    private enum class State {
        CANCEL,
        ALLOW
    }

    private class ThreadStatus {
        var mState = State.ALLOW
        var mOptions: BitmapFactory.Options? = null
        override fun toString(): String {
            var s: String = when (mState) {
                State.CANCEL -> "Cancel"
                State.ALLOW -> "Allow"
                else -> "?"
            }
            s = "thread state = $s, options = $mOptions"
            return s
        }
    }

    class ThreadSet : Iterable<Thread?> {
        private val mWeakCollection = WeakHashMap<Thread, Any?>()
        fun add(t: Thread) {
            mWeakCollection[t] = null
        }

        override fun iterator(): MutableIterator<Thread> {
            return mWeakCollection.keys.iterator()
        }
    }

    @Synchronized
    private fun getOrCreateThreadStatus(t: Thread): ThreadStatus {
        var status = mThreadStatus[t]
        if (status == null) {
            status = ThreadStatus()
            mThreadStatus[t] = status
        }
        return status
    }

    @Synchronized
    fun cancelThreadDecoding(threads: ThreadSet) {
        for (t in threads) {
            cancelThreadDecoding(t)
        }
    }

    @Synchronized
    fun cancelThreadDecoding(t: Thread) {
        val status = getOrCreateThreadStatus(t)
        status.mState = State.CANCEL
        if (status.mOptions != null) {
            status.mOptions?.requestCancelDecode()
        }
        (this as Object).notifyAll()
    }

    companion object {
        private var sManager: BitmapManager? = null
        @Synchronized
        fun instance(): BitmapManager? {
            if (sManager == null) {
                sManager = BitmapManager()
            }
            return sManager
        }
    }
}
